### ISSUE TITLE
Bug: [new config system] unmatched individual patterns are not reported

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#16275]
### Environment

Node version: v16.14.0
npm version: v8.3.1
Local ESLint version: v8.23.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.19044

### What parser are you using?

Default (Espree)

### What did you do?

I ran the following command in my local clone of the eslint/eslint repo, `main` branch:

```
node ./bin/eslint.js "tools/*.js" "doesnotexist/*.js" -f tap
```

### What did you expect to happen?

Since there is no `doesnotexist` directory, I would expect an error saying that the `doesnotexist/*.js` pattern is unmatched:

```
Oops! Something went wrong! :(

ESLint: 8.23.0

No files matching the pattern "doesnotexist/*.js" were found.
Please check for typing mistakes in the pattern.
```

This is the current behavior in eslintrc mode (when `eslint.config.js` file is deleted).

### What actually happened?

Successful linting:

```
TAP version 13
1..7
ok 1 - D:\projects\eslint\tools\code-sample-minimizer.js
ok 2 - D:\projects\eslint\tools\config-rule.js
ok 3 - D:\projects\eslint\tools\eslint-fuzzer.js
ok 4 - D:\projects\eslint\tools\fetch-docs-links.js
ok 5 - D:\projects\eslint\tools\fuzzer-runner.js
ok 6 - D:\projects\eslint\tools\update-readme.js
ok 7 - D:\projects\eslint\tools\update-rule-types.js
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

This happens because we're passing to `globby` all the patterns together, so we don't know if some of them didn't match anything.

**Comment by nzakas** [comment]
Verified as a bug.

**Comment by DevilsAutumn** [comment]
I would like to work on this issue.

**Comment by DevilsAutumn** [comment]
I tried to implement the logic of looping through globbyPatterns and checking the globbyResult (later pushing into globbyResults) for each pattern and if ```globbyResult.length == 0``` it threw the NoFilesFoundError. This logic worked perfectly fine for the above issue but while running unit tests, I encountered an error which was caused due to ```lib/config/default-conf.js``` where  ``` "**/*.mjs"```, ```"**/*.cjs"``` are intentionally added empty config to ensure these files are globbed by default. Since these files are not present in eslint repo the globbyResult came out to be empty throwing the NoFilesFoundError.

According to me there can be two solutions to this new problem:
**1**. Making empty test.mjs and test.cjs files so that globbyResult get the path to these files and prevent errors in unit tests. 
**2**. Removing  ``` "**/*.mjs"```, ```"**/*.cjs"``` from ```default-config.js``` so that they are not globbed by default.

Other solutions/suggestions are also welcomed for discussion.

**Comment by mdjermanovic** [comment]
> I tried to implement the logic of looping through globbyPatterns and checking the globbyResult (later pushing into globbyResults) for each pattern and if `globbyResult.length == 0` it threw the NoFilesFoundError. This logic worked perfectly fine for the above issue but while running unit tests, I encountered an error which was caused due to `lib/config/default-conf.js` where ` "**/*.mjs"`, `"**/*.cjs"` are intentionally added empty config to ensure these files are globbed by default. Since these files are not present in eslint repo the globbyResult came out to be empty throwing the NoFilesFoundError.

Does this happen only when the pattern is a directory path? In that case, e.g. if the directory is `foo`, we're calculating globs as `[ 'foo/**/*.js', 'foo/**/*.mjs', 'foo/**/*.cjs' ]`, but it shouldn't be required that each of them must match something. If at least one of the globs calculated from `foo` matches something, than the original argument `foo` matches something and therefore it shouldn't be reported.

**Comment by DevilsAutumn** [comment]
I think its looking for files with extension ```.mjs``` and ```.cjs``` in all the  directories inside eslint codebase,  for the unit tests  the globbyPatterns array is ```[ '**/*.js', '**/*.mjs', '**/*.cjs', 'tools/*.js', 'tests/**/*.js' ]```  and for  ```'**/*.mjs'``` and ```'**/*.cjs'``` we are not specifying directory path hence globby is returning empty array which means no related filepath is found. For ```'**/*.js'``` also we are not specifying any directory path but that is not giving error since '.js' files are present.


I am doing something like this:

```
const globbyResults = [];

    if (globbyPatterns.length) {
        for (let i = 0; i < globbyPatterns.length; i++) {
            const SingleGlobbyResult = await globby(globbyPatterns[i], {
                cwd,
                absolute: true,
                ignore: configs.ignores.filter(
                    matcher => typeof matcher === "string"
                )
            });

            if (!SingleGlobbyResult.length) {
                throw new NoFilesFoundError(globbyPatterns[i], globInputPaths);
            } else {
                globbyResults.push(...SingleGlobbyResult);
            }
        }
    }
```

**Comment by nzakas** [comment]
Thanks for taking a stab at this. It’s pretty complex logic so it’s easy to get lost in there. Definitely not a gentle landing spot for a new contributor.

@mdjermanovic is correct: we don’t want every glob pattern to throw if there’s not a match. This particular issue points out that an explicit glob pattern passed on the command line that has no match should throw an error.

I think we should have a separate discussion about how to handle when a directory is passed because this works differently than it did with eslintrc.

**Comment by nzakas** [comment]
I'll work on this now that we've completely rewritten the globbing logic.
